[SLP] Simplify tryToFindDuplicates() (NFC)

gbossu opened the pull request:
This NFC aims to simplify the control-flow and interfaces used in `tryToFindDuplicates()`. The point is to make it easier to understand where decisions for scalar de-duplication are made.

In particular:
 - Limit indentation
 - Rename some variables to better match their use case
 - Always give consistent outputs for `VL` and `ReuseShuffleIndices`. This makes it possible to use the same code for building gather `TreeEntry` everywhere. This also allows to remove the `TryToFindDuplicates` lambda.

gbossu committed:
tryToFindDuplicates(): Limit indentation

gbossu committed:
tryToFindDuplicates(): more descriptive naming

gbossu committed:
tryToFindDuplicates(): more consistent outputs

In particular, ReuseShuffleIndices is always in a consistent state,
which allows to call newTreeEntry with the same parameters regardless of
the result of tryToFindDuplicates().

gbossu commented:
FYI @alexey-bataev @RKSimon I've seen some of the refactorings made in https://github.com/llvm/llvm-project/commit/02a708b93b9ebe5e4fbc2b266da94677e6f793d3 and https://github.com/llvm/llvm-project/commit/61d04f1aac96f671a3975e04dac02d270b86ac4d and I think they are great, it's a lot easier to follow the code!

I had started some other changes already which go a bit further. They are summarized in the PR description. I've split the changes in multiple commits to make the review easier. I could also create multiple PRs if needed.  I'd be happy to hear your opinion, cheers! 🙂

gbossu commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -9967,6 +9967,12 @@ void BoUpSLP::buildTree_rec(ArrayRef<Value *> VLRef, unsigned Depth,
+    auto Invalid = ScheduleBundle::invalid();
+    newTreeEntry(VL, Invalid /*not vectorized*/, S, UserTreeIdx,
+                 ReuseShuffleIndices);
+  };
```
Now that I think about it, I'd rather get rid of this commit, or create a new method in `BoUpSLP` to replace the lambda and not implicitly capture `VL` and `ReuseShuffleIndices`:

```
  void BoUpSLP::newGatherTreeEntry(ArrayRef<Value *> VL, const InstructionsState &S,
                                   const EdgeInfo &UserTreeIdx,
                                   ArrayRef<int> ReuseShuffleIndices) {
    auto Invalid = ScheduleBundle::invalid();
    newTreeEntry(VL, Invalid, S, UserTreeIdx, ReuseShuffleIndices);
  }
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -4251,13 +4260,34 @@ class BoUpSLP {
   bool areAltOperandsProfitable(const InstructionsState &S,
                                 ArrayRef<Value *> VL) const;
 
+  /// Contains all the outputs of legality analysis for a list of values to
+  /// vectorize.
+  class ScalarsVectorizationLegality {
+    InstructionsState S;
+    bool IsLegal;
+    bool TryToFindDuplicates;
+    bool TrySplitVectorize;
+
+  public:
+    ScalarsVectorizationLegality(InstructionsState S, bool IsLegal,
+                                 bool TryToFindDuplicates = true,
+                                 bool TrySplitVectorize = false)
+        : S(S), IsLegal(IsLegal), TryToFindDuplicates(TryToFindDuplicates),
+          TrySplitVectorize(TrySplitVectorize) {
+      assert((!IsLegal || (S.valid() && TryToFindDuplicates)) &&
+             "Inconsistent state");
+    }
+    const InstructionsState &getInstructionsState() const { return S; };
+    bool isLegal() const { return IsLegal; }
+    bool tryToFindDuplicates() const { return TryToFindDuplicates; }
+    bool trySplitVectorize() const { return TrySplitVectorize; }
+  };
+
```
Move to separate patch

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -9556,55 +9590,67 @@ static bool tryToFindDuplicates(SmallVectorImpl<Value *> &VL,
-            TTI, UniqueValues.front()->getType(), UniqueValues.size());
-        PWSz = std::min<unsigned>(PWSz, VL.size());
-        if (PWSz == VL.size()) {
+    return true;
```
All these returns and code restructiong in this function better to have in a separate patch

gbossu replied:
Fine to me! I'll move the last two commits in a separate patch, and leave the first three ones here then :) They all deal with `tryToFindDuplicates()`, so I guess it does make sense to keep them together still. And it will be a pretty small patch.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -10000,31 +10043,19 @@ bool BoUpSLP::isLegalToVectorizeScalars(ArrayRef<Value *> VL, unsigned Depth,
 
   SmallVector<int> ReuseShuffleIndices;
-  SmallVector<Value *> NonUniqueValueVL(VL.begin(), VL.end());
-  auto TryToFindDuplicates = [&](const InstructionsState &S,
```
This should be in a separate patch

gbossu commented:
Thanks for the review @alexey-bataev , I have now simplified the PR so it only touches the code in `tryToFindDuplicates()`, and I'll create a follow up for refactoring the legality checks. I think it's pretty small and straight-forward now, let me know what you think :)

Note: I have also updated the PR description.
